File: lib/Game.py
# coding: latin1
from . import sfm_glob
from . import helpers
import random
import pickle
from .Player import Player
from .Team import Team
from .Division import Division
from operator import attrgetter
from . import db
from .Manager import Manager

class Game(object):
    def team_skill_per_division(self, division_id, team_id):
        total_teams = sfm_glob.COMPETITION['TOTAL_NUMBER_OF_DIVISIONS'] * sfm_glob.COMPETITION['TEAMS PER DIVISION']
        step = (sfm_glob.TEAM['MAX_SKILL'] - sfm_glob.TEAM['MIN_DIV_SKILL']) / float(total_teams)
        team_id_total = (division_id - 1) * sfm_glob.COMPETITION['TEAMS PER DIVISION'] + (team_id - 1)
        return sfm_glob.TEAM['MAX_SKILL'] - team_id_total * step

    # START OF GAME
    def start(self, human_team = None, manager = None):
        all_teams = db.TEAMS
        if human_team:
            if human_team['prev_div'] and human_team['prev_pos']:
                human_team_index = (human_team['prev_div'] - 1) * sfm_glob.COMPETITION['TEAMS PER DIVISION'] + human_team['prev_pos'] - 1
                all_teams.insert(human_team_index, human_team)

        # Create all teams
        for division_id in range(sfm_glob.COMPETITION['TOTAL_NUMBER_OF_DIVISIONS'] + 1):
            # Create division and put teams in the division
            new_division = Division(name = "League " + str(division_id + 1), level = division_id)
            teams_in_division = sfm_glob.COMPETITION['TEAMS PER DIVISION']
            if division_id == sfm_glob.COMPETITION['TOTAL_NUMBER_OF_DIVISIONS']:
                teams_in_division = sfm_glob.COMPETITION['EXTRA_TEAMS']
                new_division = Division(name = "Extra Teams", level = division_id, playable= False)

            teams = []
            for team_id in range(teams_in_division):
                name = all_teams[division_id * sfm_glob.COMPETITION['TEAMS PER DIVISION'] + team_id]['name']
                country = db.TEAMS[division_id * sfm_glob.COMPETITION['TEAMS PER DIVISION'] + team_id]['country']
                color = db.TEAMS[division_id * sfm_glob.COMPETITION['TEAMS PER DIVISION'] + team_id]['color']
                team = Team(name = name, country = country, color = color, avg_skill = self.team_skill_per_division(division_id + 1, team_id + 1), division = new_division)
                teams.append(team)

                if human_team:
                    if human_team['name'] == name:
                        prev_div = division_id
                        prev_pos = team_id
                        h_team = team

            new_division.teams = teams
            # Put division in the Game
            self.divisions.append(new_division)

        if human_team:
            self.create_human_team(h_team, prev_div, prev_pos, manager)

    def save(self, folder):
        f = open('{}/{}.sfm'.format(folder, self.name), 'wb')
        pickle.dump(self.__dict__, f)
        f.close()

    # HUMAN TEAM
    def create_human_team(self, human_team, prev_div, prev_pos, manager):
        human_team.human = True
        positions = sfm_glob.TEAM["STARTING_AMOUNT_OF_PLAYERS_PER_POS"]
        positions[random.randint(1, 2)] -= 1

        # Create players for human team
        for position, amount in enumerate(positions):
            for i in range(amount):
                if i <= (amount / 2):
                    min_skill = int(helpers.min_max(human_team.avg_skill, 1, 20))
                    max_skill = int(helpers.min_max(human_team.avg_skill + 1, 1, 20))
                else:
                    min_skill = int(helpers.min_max(human_team.avg_skill - 3, 1, 20))
                    max_skill = int(helpers.min_max(human_team.avg_skill - 1, 1, 20))
                skill = random.randint(min_skill, max_skill)

                same_country = 0.55 + 0.125 * prev_div
                country = None
                if random.random() <= same_country:
                    country = human_team.country

                human_team.players.append(Player(skill = skill, position = position, country = country, team = human_team))

        human_team.weekly_sponsorship = human_team.division.sponsorship_per_end_of_season_position(pos = min(max(prev_pos, 4), 13)) #min(max) to prevent small money in beginning
        human_team.change_finances("Sponsors", human_team.weekly_sponsorship)
        human_team.change_finances("Prize Money", human_team.division.money_per_end_of_season_position(pos = min(max(prev_pos, 4), 13))) #min(max) to prevent small money in beginning
        human_team.set_playing_tactic()
        human_team.set_transfer_list()
        human_team.order_players_by_playing_status()
        human_manager = Manager(name = manager['name'], team = human_team, human = True)
        human_manager.update_stats()
        human_team.manager = human_manager
        self.human_teams.append(human_team)
        self.managers.append(human_manager)

    def _create_ai_team(self, team, prev_div, prev_pos):
        positions = sfm_glob.TEAM["STARTING_AMOUNT_OF_PLAYERS_PER_POS"]
        team.players = []

        # Create players for human team
        for position, amount in enumerate(positions):
            for i in range(amount):
                min_skill = int(helpers.min_max(team.avg_skill - 1, 1, 20))
                max_skill = int(helpers.min_max(team.avg_skill + 1, 1, 20))

                skill = random.randint(min_skill, max_skill)

                team.players.append(Player(skill = skill, position = position, country = team.country, team = team))

        team.weekly_sponsorship = team.division.sponsorship_per_end_of_season_position(pos = prev_pos)
        team.change_finances("Sponsors", team.weekly_sponsorship)
        team.change_finances("Prize Money", team.division.money_per_end_of_season_position(pos = prev_pos))
        team.set_playing_tactic()
        team.set_transfer_list()
        team.order_players_by_playing_status()

    # START OF SEASON
    def start_of_season(self):
        self.season += 1
        self.week = 0
        for div in self.divisions:
            div.start_of_season()

        for man in self.managers:
            man.new_season()

    # END OF SEASON
    def is_season_over(self):
        if self.week >= sfm_glob.COMPETITION['TOTAL GAMES']:
            return True
        return False

    def promoted_and_demoted_teams(self):
        promoted = [[] for x in range(len(self.divisions))]
        demoted = [[] for x in range(len(self.divisions))]

        for div in self.divisions:
            if div.playable:
                div.order_table_by_position()

                promoted[div.level] = div.teams[:sfm_glob.COMPETITION['PROMOTED AND DEMOTED']]
                demoted[div.level] = div.teams[-sfm_glob.COMPETITION['PROMOTED AND DEMOTED']:]
            else:
                promoted[div.level] = self.divisions[-1].teams[:sfm_glob.COMPETITION['PROMOTED AND DEMOTED']]

        return (promoted, demoted)

    def end_of_season(self):
        def _promotions_and_demotions():
            promoted, demoted = self.promoted_and_demoted_teams()

            for div in self.divisions:
                div.order_table_by_position()
                if div.level != self.divisions[-1].level:
                    div.teams = div.teams[:-sfm_glob.COMPETITION['PROMOTED AND DEMOTED']][sfm_glob.COMPETITION['PROMOTED AND DEMOTED']:]
                else:
                    div.teams = div.teams[sfm_glob.COMPETITION['PROMOTED AND DEMOTED']:]

            for div in self.divisions:
                if div.level == 0:
                    div.teams += promoted[div.level]
                if div.level != self.divisions[-1].level:
                    div.teams += promoted[div.level + 1]
                    div.teams += demoted[div.level - 1]
                else:
                    div.teams += demoted[div.level - 1]

        def _update_team_skills():
            '''Redefines team average skills for all teams according to their positions and divisions'''
            for div in self.divisions:
                if div.playable:
                    div.order_table_by_position()
                for pos, team in enumerate(div.teams):
                    team.avg_skill = self.team_skill_per_division(div.level + 1, pos + 1)
                    if not div.playable:
                        team.avg_skill += 1

        def _is_game_over():
            if human_team in self.divisions[-1].teams:
                return True

        for div in self.divisions:
            div.end_of_season()
        self.order_divisions_by_level()
        _update_team_skills()
        _promotions_and_demotions()

        if len(self.human_teams) > 0:
            human_team = self.human_teams[0]

            if _is_game_over():
                self.ended = True


    # WEEKLY
    def next_week(self):
        def _is_game_over():
            if human_team.fan_happiness < sfm_glob.TEAM_GOALS['MIN_FAN_HAPPINESS_FOR_FIRING']:
                prob = 1 - helpers.value01(sfm_glob.TEAM_GOALS['MIN_FAN_HAPPINESS_FOR_FIRING'], sfm_glob.TEAM_GOALS["MIN_FAN_HAPPINESS"], sfm_glob.TEAM_GOALS["MAX_FAN_HAPPINESS"])
                if random.uniform(0, 1) <= prob:
                    return True
                return False

        for division in self.divisions:
            division.next_week(week = self.week)
        self.week += 1

        if len(self.human_teams) > 0:
            human_team = self.human_teams[0]

            if _is_game_over():
                self.ended = True

        for manager in self.managers:
            manager.update_stats()

    def simulate_weekly_matches(self):
        for division in self.divisions:
            division.simulate_weekly_matches(week = self.week)

    # DIVISIONS
    def order_divisions_by_level(self):
        self.divisions.sort(key=attrgetter('level'))

    # INFORMATION
    def year(self):
        return sfm_glob.GAME['STARTING YEAR'] + self.season - 1

    def __init__(self, name = None, week = None, season = None, divisions = None, human_teams = None, managers = None, career = None, last_screen = None, ended = False):
        self.name = name

        if week is None:
            week = 0
        self.week = week

        if season is None:
            season = 0
        self.season = season

        if divisions is None:
            divisions = []
        self.divisions = divisions

        if human_teams is None:
            human_teams = []
        self.human_teams = human_teams

        if managers is None:
            managers = []
        self.managers = managers

        if last_screen is None:
            last_screen = "MainScreen"
        self.last_screen = last_screen

        self.ended = ended

File: webapp/src/game/game.ts
import { COMPETITION, GAME, TEAM, TEAM_GOALS } from './constants';
import { TEAMS as DB_TEAMS } from './db';
import { minMax, normalize, randomInt, value01 } from './helpers';
import { Division } from './division';
import { Manager } from './manager';
import { Player } from './player';
import { Team as TeamEntity } from './team';

interface HumanTeamConfig {
  name: string;
  color: string;
  country: string;
  prev_pos?: number | null;
  prev_div?: number | null;
}

interface ManagerConfig {
  name: string;
}

export class Game {
  divisions: Division[];
  humanTeams: TeamEntity[];
  managers: Manager[];
  lastScreen: string;
  ended: boolean;

  constructor(
    public name: string,
    public week = 0,
    public season = 0,
    divisions: Division[] = [],
    humanTeams: TeamEntity[] = [],
    managers: Manager[] = [],
    lastScreen = 'MainScreen',
    ended = false
  ) {
    this.divisions = divisions;
    this.humanTeams = humanTeams;
    this.managers = managers;
    this.lastScreen = lastScreen;
    this.ended = ended;
  }

  teamSkillPerDivision(divisionId: number, teamId: number): number {
    const totalTeams = COMPETITION['TOTAL_NUMBER_OF_DIVISIONS'] * COMPETITION['TEAMS PER DIVISION'];
    const step = (TEAM['MAX_SKILL'] - TEAM['MIN_DIV_SKILL']) / totalTeams;
    const teamIndex = (divisionId - 1) * COMPETITION['TEAMS PER DIVISION'] + (teamId - 1);
    return TEAM['MAX_SKILL'] - teamIndex * step;
  }

  start(humanTeam?: HumanTeamConfig, managerConfig?: ManagerConfig): void {
    const allTeams = [...DB_TEAMS];
    let humanTeamEntry: HumanTeamConfig | null = humanTeam ?? null;

    if (humanTeamEntry && humanTeamEntry.prev_div && humanTeamEntry.prev_pos) {
      const index =
        (humanTeamEntry.prev_div - 1) * COMPETITION['TEAMS PER DIVISION'] + (humanTeamEntry.prev_pos - 1);
      allTeams.splice(index, 0, {
        name: humanTeamEntry.name,
        country: humanTeamEntry.country,
        color: humanTeamEntry.color
      });
    }

    this.divisions = [];
    for (let divisionId = 0; divisionId <= COMPETITION['TOTAL_NUMBER_OF_DIVISIONS']; divisionId += 1) {
      const isExtra = divisionId === COMPETITION['TOTAL_NUMBER_OF_DIVISIONS'];
      const levelName = isExtra ? 'Extra Teams' : `League ${divisionId + 1}`;
      const division = new Division(levelName, divisionId, [], [], !isExtra);
      const teamsInDivision = isExtra ? COMPETITION['EXTRA_TEAMS'] : COMPETITION['TEAMS PER DIVISION'];
      const teams: TeamEntity[] = [];

      for (let teamId = 0; teamId < teamsInDivision; teamId += 1) {
        const teamInfo = allTeams[divisionId * COMPETITION['TEAMS PER DIVISION'] + teamId];
        if (!teamInfo) {
          continue;
        }
        const avgSkill = this.teamSkillPerDivision(divisionId + 1, teamId + 1);
        const team = new TeamEntity(teamInfo.name, teamInfo.country, teamInfo.color, {
          avgSkill,
          division
        });
        teams.push(team);
        if (humanTeamEntry && teamInfo.name === humanTeamEntry.name) {
          humanTeamEntry = { ...humanTeamEntry, prev_div: divisionId, prev_pos: teamId };
        }
      }

      division.teams = teams;
      this.divisions.push(division);
    }

    if (humanTeamEntry && managerConfig) {
      const divisionIndex = humanTeamEntry.prev_div ?? 0;
      const teamIndex = humanTeamEntry.prev_pos ?? 0;
      const team = this.divisions[divisionIndex].teams[teamIndex];
      this.createHumanTeam(team, divisionIndex, teamIndex, managerConfig);
    }
  }

  private createHumanTeam(
    humanTeam: TeamEntity,
    prevDiv: number,
    prevPos: number,
    managerConfig: ManagerConfig
  ): void {
    humanTeam.human = true;
    const positions = [...TEAM['STARTING_AMOUNT_OF_PLAYERS_PER_POS']];
    positions[randomInt(1, 2)] -= 1;

    for (let position = 0; position < positions.length; position += 1) {
      for (let i = 0; i < positions[position]; i += 1) {
        let minSkill: number;
        let maxSkill: number;
        if (i <= positions[position] / 2) {
          minSkill = minMax(humanTeam.avgSkill, 1, 20);
          maxSkill = minMax(humanTeam.avgSkill + 1, 1, 20);
        } else {
          minSkill = minMax(humanTeam.avgSkill - 3, 1, 20);
          maxSkill = minMax(humanTeam.avgSkill - 1, 1, 20);
        }
        const skill = randomInt(Math.floor(minSkill), Math.floor(maxSkill));
        const sameCountry = 0.55 + 0.125 * prevDiv;
        const country = Math.random() <= sameCountry ? humanTeam.country : null;
        const player = new Player({ skill, position, country, team: humanTeam });
        humanTeam.players.push(player);
      }
    }

    const sponsorshipPos = Math.min(Math.max(prevPos, 4), 13);
    humanTeam.weeklySponsorship = humanTeam.division?.sponsorshipPerEndOfSeasonPosition(sponsorshipPos) ?? 0;
    humanTeam.changeFinances('Sponsors', humanTeam.weeklySponsorship);
    humanTeam.changeFinances(
      'Prize Money',
      humanTeam.division?.moneyPerEndOfSeasonPosition(sponsorshipPos) ?? 0
    );
    humanTeam.setPlayingTactic();
    humanTeam.setTransferList();
    humanTeam.orderPlayersByPlayingStatus();

    const manager = new Manager(managerConfig.name, humanTeam, true);
    manager.updateStats();
    humanTeam.manager = manager;
    this.humanTeams = [humanTeam];
    this.managers.push(manager);
  }

  startOfSeason(): void {
    this.season += 1;
    this.week = 0;
    this.divisions.forEach((division) => division.startOfSeason());
    this.managers.forEach((manager) => manager.newSeason());
  }

  isSeasonOver(): boolean {
    return this.week >= COMPETITION['TOTAL GAMES'];
  }

  promotedAndDemotedTeams(): { promoted: TeamEntity[][]; demoted: TeamEntity[][] } {
    const promoted: TeamEntity[][] = Array.from({ length: this.divisions.length }, () => []);
    const demoted: TeamEntity[][] = Array.from({ length: this.divisions.length }, () => []);

    this.divisions.forEach((division) => {
      if (division.playable) {
        division.orderTableByPosition();
        promoted[division.level] = division.teams.slice(0, COMPETITION['PROMOTED AND DEMOTED']);
        demoted[division.level] = division.teams.slice(-COMPETITION['PROMOTED AND DEMOTED']);
      } else {
        promoted[division.level] = this.divisions[this.divisions.length - 1].teams.slice(
          0,
          COMPETITION['PROMOTED AND DEMOTED']
        );
      }
    });

    return { promoted, demoted };
  }

  endOfSeason(): void {
    this.divisions.forEach((division) => division.endOfSeason());
    this.orderDivisionsByLevel();

    const { promoted, demoted } = this.promotedAndDemotedTeams();

    this.divisions.forEach((division) => {
      division.orderTableByPosition();
      if (division.level !== this.divisions[this.divisions.length - 1].level) {
        division.teams = division.teams.slice(
          COMPETITION['PROMOTED AND DEMOTED'],
          division.teams.length - COMPETITION['PROMOTED AND DEMOTED']
        );
      } else {
        division.teams = division.teams.slice(COMPETITION['PROMOTED AND DEMOTED']);
      }
    });

    this.divisions.forEach((division, index) => {
      if (division.level === 0) {
        division.teams = [...division.teams, ...promoted[index]];
      }
      if (division.level !== this.divisions[this.divisions.length - 1].level) {
        const promotedNext = promoted[index + 1] ?? [];
        const demotedPrev = demoted[(index - 1 + this.divisions.length) % this.divisions.length] ?? [];
        division.teams = [...division.teams, ...promotedNext, ...demotedPrev];
      } else {
        const demotedPrev = demoted[(index - 1 + this.divisions.length) % this.divisions.length] ?? [];
        division.teams = [...division.teams, ...demotedPrev];
      }
    });

    this.divisions.forEach((division) => {
      division.teams.forEach((team, pos) => {
        team.avgSkill = this.teamSkillPerDivision(division.level + 1, pos + 1);
        if (!division.playable) {
          team.avgSkill += 1;
        }
      });
    });

    if (this.humanTeams.length > 0) {
      const humanTeam = this.humanTeams[0];
      const extraDivision = this.divisions[this.divisions.length - 1];
      if (extraDivision.teams.includes(humanTeam)) {
        this.ended = true;
      }
    }
  }

  nextWeek(): void {
    this.divisions.forEach((division) => division.nextWeek(this.week));
    this.week += 1;

    if (this.humanTeams.length > 0) {
      const humanTeam = this.humanTeams[0];
      if (humanTeam.fanHappiness < TEAM_GOALS['MIN_FAN_HAPPINESS_FOR_FIRING']) {
        const prob = 1 - value01(
          TEAM_GOALS['MIN_FAN_HAPPINESS_FOR_FIRING'],
          TEAM_GOALS['MIN_FAN_HAPPINESS'],
          TEAM_GOALS['MAX_FAN_HAPPINESS']
        );
        if (Math.random() <= prob) {
          this.ended = true;
        }
      }
    }

    this.managers.forEach((manager) => manager.updateStats());
  }

  simulateWeeklyMatches(): void {
    this.divisions.forEach((division) => division.simulateWeeklyMatches(this.week));
  }

  orderDivisionsByLevel(): void {
    this.divisions.sort((a, b) => a.level - b.level);
  }

  year(): number {
    return GAME['STARTING YEAR'] + this.season - 1;
  }

  serialize(): object {
    return {
      name: this.name,
      week: this.week,
      season: this.season,
      ended: this.ended,
      lastScreen: this.lastScreen
    };
  }
}
